File: Modules/CreateBatchFiles.m
function handles = CreateBatchFiles(handles)

% Help for the Create Batch Files module:
% Category: File Processing
%
% SHORT DESCRIPTION:
% Produces text files which allow individual batches of images to be
% processed separately on a cluster of computers.
% *************************************************************************
%
% This module creates a set of files that can be submitted in parallel to a
% cluster for faster processing. This module should be placed at the end of
% an image processing pipeline.
%
% Before using this module, you should read Help -> General Help ->
% Batch Processing. That help file also will instruct you on how to
% run the batch files that are created by this module.
%
% Settings:
% PATH TO BATCH_DATA.MAT FILE:
% Choose either the Default Output Folder (".") or a path relative to it,
%   e.g. "./SaveToThisFolder"
%
% Other Paths: The last two settings allow changing the paths between
% local and cluster computers. For example, when starting with a local PC
% computer and setting a batch to run on a Linux-based cluster, 
% the path may be essentially the same except the first notation:
%
% PC:    \\remoteserver1\cluster\project
% Linux: /remoteserver2/cluster/project
%
% In this case, for the local machine you would type "\\remoteserver1" and
% for the remote machine you would type "/remoteserver2". As of now, this
% is hardcoded to always use Linux and Macintosh style slashes (/).
%
% If your input image folder and output folder are located on different 
% machines, you can specify the input image paths followed by the output 
% path separated by a comma.
%
% Note: This module produces a Batch_data.mat file. This contains the
% first image set's measurements plus information about the processing
% that each batch file needs access to in order to initialize the
% processing.  See the BatchRunner.py, CPCluster.py, and CPCluster.m
% files for how this information is used.  As many clusters use
% different configurations for batch control, compiled versus
% interpreted Matlab, access paths, etc. it will probably be necessary
% to use those files as guides for a locally customized solution.
% BatchRunner.py requires Python 2.5.2 and the module scipy 0.6 to be installed. 
%
% See also MergeOutputFiles, GSBatchProcessing.

% CellProfiler is distributed under the GNU General Public License.
% See the accompanying file LICENSE for details.
%
% Developed by the Whitehead Institute for Biomedical Research.
% Copyright 2003,2004,2005.
%
% Please see the AUTHORS file for credits.
%
% Website: http://www.cellprofiler.org
%
% $Revision$

% Notes for pyCP:
% The First setting stays the same, except that perhaps the "Use . to specify
% default" is replaced with some GUI element that lets you choose "Use
% default output folder" from a dropdown menu rather than typing an arbitrary period.
%
% The second and third settings are translations from local-to-cluster nomenclature.
% They are notorious for tripping up users, so clarification would be
% useful. First thing is that we should ask whether the local and cluster
% machines use the same pathnames (or something like that) and skip the
% rest if the answer is Yes.
% The questions have 1 or 2 inputs, for default image and ouput folders, but this 
% should be more explicit.  And it's possible that more than 2 translations 
% could be necessary (if images were on two different servers and the output was
% on a third), so there should be no limitation on the number of translations.  
% It would be nice if initially two boxes were presented
% on the left and right that were labeled LOCAL ROOT PATH and CLUSTER ROOT PATH.  A button 
% labeled "Do you want to add another set of paths?", if clicked, would display another pair 
% of text boxes below the first set, to be filled in.  
% Further, it would be good (if possible) to show an example fully-constructed 
% path as the user types in the root path. 
% Maybe pie-in-the-sky, but a test mode would be really cool, 
% to make a simple (webserver?) call to see if the cluster's path existed, 
% which would head off a lot of errors.  We should also prevent PC users from 
% entering UNC paths ('\\...') and stick to mapped drives (or pick one), since 
% we don't want to map all possible network naming schemes. 

%%%%%%%%%%%%%%%%%
%%% VARIABLES %%%
%%%%%%%%%%%%%%%%%
drawnow

[CurrentModule, CurrentModuleNum, ModuleName] = CPwhichmodule(handles);

%pathnametextVAR01 = What is the path to the folder where the batch control file (Batch_data.mat) will be saved? Leave a period (.) to use the default output folder.
%defaultVAR01 = .
BatchSavePath = char(handles.Settings.VariableValues{CurrentModuleNum,1});

%pathnametextVAR02 = LOCAL MACHINE'S root path (omitting trailing slashes). Leave a period (.) if LOCAL and CLUSTER root paths are the same. If your image and output folder paths are different, enter the input then the output path separated by a comma.
%defaultVAR02 = .
OldPathname = char(handles.Settings.VariableValues{CurrentModuleNum,2});

%pathnametextVAR03 = CLUSTER MACHINE'S root path (omitting trailing slashes). Leave a period (.) if LOCAL and CLUSTER root paths are the same. See above for entering multiple paths.
%defaultVAR03 = .
NewPathname = char(handles.Settings.VariableValues{CurrentModuleNum,3});

%textVAR04 = Note: This module must be the last one in the analysis pipeline.

%%%VariableRevisionNumber = 8

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% PRELIMINARY CALCULATIONS & FILE HANDLING %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
drawnow

% If this isn't the first cycle, we are running on the
% cluster, and should just continue.
if (handles.Current.SetBeingAnalyzed > 1) || isfield(handles.Current, 'BatchInfo')
    return;
end

isImageGroups = isfield(handles.Pipeline,'ImageGroupFields');
if ~isImageGroups
    if handles.Current.NumberOfImageSets == 1 
       CPwarndlg(['Warning: No batch scripts have been written because ',...
           'you have scheduled only one cycle to be processed and that cycle is already complete.']);
       return;
    end  
end

if strncmp(BatchSavePath, '.',1)
    if length(BatchSavePath) == 1
        BatchSavePath = handles.Current.DefaultOutputDirectory;
    else
        BatchSavePath = fullfile(handles.Current.DefaultOutputDirectory,BatchSavePath(2:end));
    end
end

% Check that Batch_data.mat does not already exist
PathAndFileName = fullfile(BatchSavePath, 'Batch_data.mat');
if exist(PathAndFileName,'file') == 2
    button = CPquestdlg('Batch_data.mat already exists in the output directory.  Are you sure you want to overwrite Batch_data.mat?',...
        'Overwrite Batch_data');
    if ~strcmp(button,'Yes')
        
        set(handles.timertexthandle,'string','Canceling after current module')
        CPmsgbox(['Image processing was canceled in the ', ModuleName, ' module at your request.'])
        CPclosefigure(handles,CurrentModule)
        return

    end
end


% Checks that this is the last module in the analysis path.
if (CurrentModuleNum ~= handles.Current.NumberOfModules),
    error(['Image processing was canceled because ', ModuleName, ' must be the last module in the pipeline.']);
end

% Saves a copy of the handles structure to revert back to later. The
% altered handles must be saved using the variable name 'handles'
% because the save function will not allow us to save a variable
% under a different name.
PreservedHandles = handles;
 
% Changes parts of several pathnames if the user has
% specified that parts of the pathname are named differently from
% the perspective of the local computer vs. the cluster
% machines

% Parse OldPathname, checking for comma separating multiple paths
if ~strcmp(OldPathname,strtok(OldPathname,',')),
    [firstpathstr,secondpathstr] = strtok(OldPathname,',');
    s = cell(1,2); s{1} = strtrim(firstpathstr); s{2} = strtrim(secondpathstr(2:end));
    OldPathname = s;
    if ~strcmp(NewPathname,strtok(NewPathname,',')),
        [firstpathstr,secondpathstr] = strtok(NewPathname,',');
        s = cell(1,2); s{1} = strtrim(firstpathstr); s{2} = strtrim(secondpathstr(2:end));
        NewPathname = s;
    else
        error(['Processing was canceled in the ', ModuleName, 'module because if there are two pathnames specified for the local machine, there need to be two pathnames for the cluster machine.']);
    end
else    % If no comma, one path only. Duplicate to both strings
    s = cell(1,2); [s{:}] = deal(strtrim(OldPathname));
    OldPathname = s;
    s = cell(1,2); [s{:}] = deal(strtrim(NewPathname));
    NewPathname = s;
end

if ~any(strcmp(OldPathname, '.'))
    % Changes the default output and image pathnames
    for i = 1:length(OldPathname)
		NewDefaultImageDirectory = strrep(strrep(handles.Current.DefaultImageDirectory,OldPathname{i},NewPathname{i}),'\','/');
		handles.Current.DefaultImageDirectory = NewDefaultImageDirectory;
		
		NewDefaultOutputDirectory = strrep(strrep(handles.Current.DefaultOutputDirectory,OldPathname{i},NewPathname{i}),'\','/');
		handles.Current.DefaultOutputDirectory = NewDefaultOutputDirectory;
	end
    
    % Replaces \ with / in all image filenames (only relevant for PCs)
    % (a) handles.Pipelines
    Fieldnames = fieldnames(handles.Pipeline);
    FileListFieldNames = Fieldnames(strncmp(Fieldnames, 'FileList', 8));
    for i = 1:length(FileListFieldNames)
        if ndims(handles.Pipeline.(FileListFieldNames{i})) == 2
            % Assumed to be FLEX file, with filename in the 1st row of
            % cell array
            for j = 1:size(handles.Pipeline.(FileListFieldNames{i}),2)
                handles.Pipeline.(FileListFieldNames{i}){1,j} = strrep(handles.Pipeline.(FileListFieldNames{i}){1,j},'\','/');
            end
        else
            handles.Pipeline.(FileListFieldNames{i}) = strrep(handles.Pipeline.(FileListFieldNames{i}),'\','/');
        end
    end
    % (b) handles.Pipelines.GroupFileList (if it exists)
    if isfield(handles.Pipeline,'ImageGroupFields')
        for i = 1:length(handles.Pipeline.GroupFileList)
            Fieldnames = fieldnames(handles.Pipeline.GroupFileList{i});
            FileListFieldNames = Fieldnames(strncmp(Fieldnames, 'FileList', 8));
            for j = 1:length(FileListFieldNames)
                if ndims(handles.Pipeline.GroupFileList{i}.(FileListFieldNames{j})) == 2
                    % Assumed to be FLEX file, with filename in the 1st row of
                    % cell array
                    for k = 1:size(handles.Pipeline.GroupFileList{i}.(FileListFieldNames{j}),2)
                        handles.Pipeline.GroupFileList{i}.(FileListFieldNames{j}){1,k} = strrep(handles.Pipeline.GroupFileList{i}.(FileListFieldNames{j}){1,k},'\','/');
                    end
                else
                    handles.Pipeline.GroupFileList{i}.(FileListFieldNames{j}) = strrep(handles.Pipeline.GroupFileList{i}.(FileListFieldNames{j}),'\','/');
                end
            end
        end
    end
    
    % Deal with input paths that have already been saved to the handles
    % (a) handles.Pipelines
    for k = 1:length(OldPathname)
		Fieldnames = fieldnames(handles.Pipeline);
		PathFieldnames = Fieldnames(strncmpi(Fieldnames,'pathname',8));
		for i = 1:length(PathFieldnames),
			handles.Pipeline.(PathFieldnames{i}) = strrep(strrep(handles.Pipeline.(PathFieldnames{i}),OldPathname{k},NewPathname{k}),'\','/');
			handles.Pipeline.(PathFieldnames{i}) = strrep(strrep(handles.Pipeline.(PathFieldnames{i}),OldPathname{k},NewPathname{k}),'\','/');
		end
		% (b) handles.Pipelines.GroupFileList (if it exists)
		if isfield(handles.Pipeline,'ImageGroupFields')
			for i = 1:length(handles.Pipeline.GroupFileList)
				Fieldnames = fieldnames(handles.Pipeline.GroupFileList{i});
				PathFieldnames = Fieldnames(strncmpi(Fieldnames,'pathname',8));
				for j = 1:length(PathFieldnames),
					handles.Pipeline.GroupFileList{i}.(PathFieldnames{j}) = strrep(strrep(handles.Pipeline.GroupFileList{i}.(PathFieldnames{j}),OldPathname{k},NewPathname{k}),'\','/');
					handles.Pipeline.GroupFileList{i}.(PathFieldnames{j}) = strrep(strrep(handles.Pipeline.GroupFileList{i}.(PathFieldnames{j}),OldPathname{k},NewPathname{k}),'\','/');
				end
			end
		end
		% (c) handles.Measurements.Image
		Fieldnames = fieldnames(handles.Measurements.Image);
		PathFieldnames = Fieldnames(strncmpi(Fieldnames,'pathname',8));
		for i = 1:length(PathFieldnames),
			handles.Measurements.Image.(PathFieldnames{i}) = strrep(strrep(handles.Measurements.Image.(PathFieldnames{i}),OldPathname{k},NewPathname{k}),'\','/');
			handles.Measurements.Image.(PathFieldnames{i}) = strrep(strrep(handles.Measurements.Image.(PathFieldnames{i}),OldPathname{k},NewPathname{k}),'\','/');
		end
	end
end

% % Make sure ModuleError has same number of elements as
% % ModuleErrorFeatures
% handles.Measurements.Image.ModuleError{handles.Current.SetBeingAnalyzed}(1,CurrentModuleNum) = 0;
% I'm not entirely sure why we need to record that there were no module
% errors in the CreateBatchFiles module here rather than in
% CellProfiler.m, but will go ahead and update the old code to the new
% CPaddmeasurements format:
handles = CPaddmeasurements(handles,'Image',CPjoinstrings('ModuleError',[CPtwodigitstring(CurrentModuleNum),ModuleName]),0);

% Python can't load function pointers
handles = rmfield(handles, 'FunctionHandles');

% The website just needs the # of image sets, so we make a global
% variable here to save
number_of_image_sets = handles.Current.NumberOfImageSets; %#ok<NASGU>

% Saves the altered handles in a file which the user will feed to
% the remote machines.
save(PathAndFileName, 'handles','number_of_image_sets');

% Reverts to the preserved handles.  (Probably not necessary, but simpler.)
handles = PreservedHandles;

CPhelpdlg('Batch files have been written.  This analysis pipeline will now stop.  You should submit the batch files for processing on your cluster. See Help > General Help > BatchProcessing for more information.', 'BatchFilesDialog');

% This is the first cycle, so this is the first time seeing this
% module.  It should cause a cancel so no further processing is done
% on this machine.
set(handles.timertexthandle,'string','Cancel')

%%%%%%%%%%%%%%%%%%%%%%%
%%% DISPLAY RESULTS %%%
%%%%%%%%%%%%%%%%%%%%%%%
drawnow

% The figure window display is unnecessary for this module, so it is
% closed during the starting image cycle.
CPclosefigure(handles,CurrentModule)

File: pyCellProfiler/cellprofiler/modules/createbatchfiles.py
'''<b>Create Batch Files</b> produces files that allow individual batches of images to be processed
separately on a cluster of computers.
<hr>
This module creates a file that can be submitted in parallel to a
cluster for faster processing. This module should be placed at the end of
an image processing pipeline.

If your computer mounts the file system differently than the cluster computers,
<b>CreateBatchFiles</b> can replace the neccesary parts of the paths to the 
image and output files. For instance, at the Broad, our Windows machines may 
access files images by mounting the file system using a drive letter, like this:
<i>Z:\imaging_analysis</i>
and our cluster computers access the same file system like this:
<i>/imaging/analysis</i>
In this case, the local root path is <i>Z:\imaging_analysis</i> and the cluster 
root path is <i>/imaging/analysis</i>.
'''
#CellProfiler is distributed under the GNU General Public License.
#See the accompanying file LICENSE for details.
#
#Developed by the Broad Institute
#Copyright 2003-2009
#
#Please see the AUTHORS file for credits.
#
#Website: http://www.cellprofiler.org

__version__="$Revision$"

import numpy as np
import os
import re
import sys
import wx
import zlib

import cellprofiler.cpimage as cpi
import cellprofiler.cpmodule as cpm
import cellprofiler.pipeline as cpp
import cellprofiler.settings as cps
import cellprofiler.preferences as cpprefs

'''# of settings aside from the mappings'''
S_FIXED_COUNT = 6
'''# of settings per mapping'''
S_PER_MAPPING = 2

'''Name of the batch data file'''
F_BATCH_DATA = 'Batch_data.mat'

class CreateBatchFiles(cpm.CPModule):
    #
    # How it works:
    #
    # There are two hidden settings: batch_mode and pickled_image_set_list
    # batch_mode controls the mode: False means "save the pipeline" and
    #     True means "run the pipeline"
    # pickled_image_set_list holds the state of the image set list. If
    #     batch_mode is False, we save the state of the image set list in
    #     pickled_image_set_list. If batch_mode is True, we load the state
    #     from pickled_image_set_list.
    #
    module_name = "CreateBatchFiles"
    category = 'File Processing'
    variable_revision_number = 4
    
    #
    def create_settings(self):
        '''Create the module settings and name the module'''
        self.wants_default_output_directory = cps.Binary("Store batch files in default output directory", True,doc="""
                Do you want to store the batch files in the default output directory? 
                Check this box to store batch files in the Default Output directory. Uncheck
                the box to enter the path to the directory that will be used to store
                these files.""")
        
        self.custom_output_directory = cps.Text("Output directory path",
                                                cpprefs.get_default_output_directory(),doc="""
                                                What is the path to the output directory?""")
        
        # Worded this way not because I am windows-centric but because it's
        # easier than listing every other OS in the universe except for VMS
        self.remote_host_is_windows = cps.Binary("Are the cluster computers running Windows?",
                                                 False,doc="""
                Check this box if the cluster computers are running one of the Microsoft
                Windows operating systems. If you check this box, <b>CreateBatchFiles</b> will
                modify all paths to use the Windows file separator (backslash, &#92;). If you
                leave the box unchecked, <b>CreateBatchFiles</b> will modify all paths to use
                the Unix or Macintosh file separator (slash,&#47;).""")
        
        self.batch_mode = cps.Binary("Hidden: in batch mode", False)
        self.default_image_directory = cps.Setting("Hidden: default image directory at time of save",
                                                   cpprefs.get_default_image_directory())
        self.revision = cps.Integer("Hidden: SVN revision number", 0)
        self.mappings = []
        self.add_mapping()
        self.add_mapping_button = cps.DoSomething("Add another path?","Add",
                                                  self.add_mapping, doc="""
                Press this button to add another path mapping.""")
    
    def add_mapping(self):
        group = cps.SettingsGroup()
        group.append("local_directory",
                     cps.Text("Local root path",
                                cpprefs.get_default_image_directory(),doc="""
                                What is the path to files on this computer? 
                                This is the root path on the local machine (i.e., the computer setting up
                                the batch file). If <b>CreateBatchFiles</b> finds
                                any pathname that matches the local root path at the begining, it will replace the
                                start with the cluster root path.
                                <p>For example, if you have mapped the remote cluster machine as:<br><br>
                                <i>Z:\your_data\images</i> (on a Windows machine, for instance)<br><br>
                                and the cluster machine sees the same directory as:<br><br>
                                <i>/server_name/your_name/your_data/images</i><br><br>
                                you would want to put <i>Z:</i> here and <i>/server_name/your_name/</i> 
                                for the cluster path in the next setting."""))

        group.append("remote_directory",
                     cps.Text("Cluster root path",
                                cpprefs.get_default_image_directory(),doc="""
                                What is the path to files on the cluster? This is the cluster 
                                root path, i.e, how the cluster machine sees the
                                top-most directory where your input/output files are stored.
                                <p>For example, if you have mapped the remote cluster machine as:<br><br>
                                <i>Z:\your_data\images</i> (on a Windows machine, for instance)<br><br>
                                and the cluster machine sees the same directory as:<br><br>
                                <i>/server_name/your_name/your_data/images</i><br><br>
                                you would want to put <i>Z:</i> in the previous setting for the
                                local machine path and <i>/server_name/your_name/</i> here. """))
        group.append("remover",
                     cps.RemoveSettingButton("", "Remove above mapping", self.mappings, group))
        group.append("divider", cps.Divider(line=False))
        self.mappings.append(group)

    def settings(self):
        result = [self.wants_default_output_directory,
                  self.custom_output_directory, self.remote_host_is_windows,
                  self.batch_mode, 
                  self.default_image_directory, self.revision]
        for mapping in self.mappings:
            result += [mapping.local_directory, mapping.remote_directory]
        return result
    
    def prepare_settings(self, setting_values):
        if (len(setting_values) - S_FIXED_COUNT) % S_PER_MAPPING != 0:
            raise ValueError("# of mapping settings (%d) "
                             "is not a multiple of %d" %
                             (len(setting_values) - S_FIXED_COUNT, 
                             S_PER_MAPPING))
        mapping_count = (len(setting_values) - S_FIXED_COUNT) / S_PER_MAPPING
        while mapping_count < len(self.mappings):
            del self.mappings[-1]
        
        while mapping_count > len(self.mappings):
            self.add_mapping()
    
    def visible_settings(self):
        result = [self.wants_default_output_directory]
        if not self.wants_default_output_directory.value:
            result += [self.custom_output_directory]
        result += [self.remote_host_is_windows]
        for mapping in self.mappings:
            result += mapping.unpack_group()
        result += [self.add_mapping_button]
        return result
    
    def prepare_run(self, pipeline, image_set_list, frame):
        '''Invoke the image_set_list pickling mechanism and save the pipeline'''
        if self.batch_mode.value:
            self.enter_batch_mode(pipeline, image_set_list)
            return True
        else:
            self.save_pipeline(pipeline, image_set_list, frame)
            return False
    
    def run(self, workspace):
        # all the actual work is done in prepare_run
        pass

    def validate_module(self, pipeline):
        '''Make sure this is the last module in the pipeline'''
        if self.module_num != len(pipeline.modules()):
            raise cps.ValidationError("The CreateBatchFiles module must be "
                                      "the last in the pipeline.",
                                      self.wants_default_output_directory)
    
    def save_pipeline(self, pipeline, image_set_list, frame):
        '''Save the pipeline in Batch_data.mat
        
        Save the pickled image_set_list state in a setting and put this
        module in batch mode.
        '''
        from cellprofiler.utilities.get_revision import version
        assert isinstance(image_set_list, cpi.ImageSetList)
        assert isinstance(pipeline, cpp.Pipeline)
        pipeline = pipeline.copy()
        pipeline.prepare_to_create_batch(image_set_list, self.alter_path)
        bizarro_self = pipeline.module(self.module_num)
        assert isinstance(bizarro_self, CreateBatchFiles)
        state = image_set_list.save_state()
        state = zlib.compress(state)
        bizarro_self.revision.value = version
        bizarro_self.batch_state = np.fromstring(state, np.uint8)
        if self.wants_default_output_directory:
            bizarro_self.custom_output_directory.value = \
                        self.alter_path(cpprefs.get_default_output_directory())
        bizarro_self.default_image_directory.value = \
                    self.alter_path(cpprefs.get_default_image_directory())
        bizarro_self.batch_mode.value = True
        if self.wants_default_output_directory.value:
            path = cpprefs.get_default_output_directory()
        else:
            path = cpprefs.get_absolute_path(self.custom_output_directory.value)
        path = os.path.join(path, F_BATCH_DATA)
        if os.path.exists(path) and frame is not None:
            if (wx.MessageBox("%s already exists. Do you want to overwrite it?"%
                              path,
                              "Overwriting %s" % F_BATCH_DATA,
                              wx.YES_NO, frame) == wx.ID_NO):
                return
        
        pipeline.save(path)
    
    def in_batch_mode(self):
        '''Tell the system whether we are in batch mode on the cluster'''
        return self.batch_mode.value
    
    def enter_batch_mode(self, pipeline, image_set_list):
        '''Restore the image set list from its setting as we go into batch mode'''
        assert isinstance(image_set_list, cpi.ImageSetList)
        assert isinstance(pipeline, cpp.Pipeline)
        state = zlib.decompress(self.batch_state.tostring())
        image_set_list.load_state(state)
        cpprefs.set_default_output_directory(self.custom_output_directory.value)
        cpprefs.set_default_image_directory(self.default_image_directory.value)
    
    def turn_off_batch_mode(self):
        '''Remove any indications that we are in batch mode
        
        This call restores the module to an editable state.
        '''
        self.batch_mode.value = False
        self.batch_state = np.zeros((0,),np.uint8)
    
    def alter_path(self, path, **varargs):
        '''Modify the path passed so that it can be executed on the remote host
        
        path = path to modify
        regexp_substitution - if true, exclude \g<...> from substitution
        '''
        for mapping in self.mappings:
            if sys.platform.startswith('win'):
                # Windows is case-insentitve so do case-insensitve mapping
                if path.upper().startswith(mapping.local_directory.value.upper()):
                    path = (mapping.remote_directory.value +
                            path[len(mapping.local_directory.value):])
            else:
                if path.startswith(mapping.local_directory.value):
                    path = (mapping.remote_directory.value +
                            path[len(mapping.local_directory.value):])
        if self.remote_host_is_windows.value:
            path = path.replace('/','\\')
        elif (varargs.has_key("regexp_substitution") and
                 varargs["regexp_substitution"]):
            path = re.subn('\\\\\\\\','/',path)[0]
            path = re.subn('\\\\(?!g\\<[^>]*\\>)','/',path)[0]
        else:
            path = path.replace('\\','/')
        return path
        
            
    def upgrade_settings(self, setting_values, variable_revision_number,
                         module_name, from_matlab):
        if from_matlab and variable_revision_number == 8:
            batch_save_path, old_pathname, new_pathname = setting_values[:3]
            if batch_save_path == '.':
                wants_default_output_directory = cps.YES
                batch_save_path = cpprefs.get_default_output_directory()
            else:
                wants_default_output_directory = cps.NO
            old_pathnames = old_pathname.split(',')
            new_pathnames = new_pathname.split(',')
            if len(old_pathnames) != len(new_pathnames):
                raise ValueError("Number of pathnames does not match. "
                                 "%d local pathnames, but %d remote pathnames" %
                                 (len(old_pathnames), len(new_pathnames)))
            setting_values = [wants_default_output_directory, batch_save_path,
                              cps.NO, cps.NO, ""]
            for old_pathname, new_pathname in zip(old_pathnames, new_pathnames):
                setting_values += [old_pathname, new_pathname]
            from_matlab = False
            variable_revision_number = 1
        if (not from_matlab) and variable_revision_number == 1:
            setting_values = (setting_values[:5] + 
                              [cpprefs.get_default_image_directory()] +
                              setting_values[5:])
            variable_revision_number = 2
        if (not from_matlab) and variable_revision_number == 2:
            from cellprofiler.utilities.get_revision import version
            
            setting_values = (setting_values[:6] + 
                              [version] +
                              setting_values[6:])
            variable_revision_number = 3
        if (not from_matlab) and variable_revision_number == 3:
            # Pickled image list is now the batch state
            self.batch_state = np.fromstring(zlib.compress(setting_values[4]),
                                             np.uint8)
            setting_values = setting_values[:4]+setting_values[5:]
            variable_revision_number = 4
        return setting_values, variable_revision_number, from_matlab
    
